### PR TITLE
Improving Expression Support for In Database

### LINKED ISSUE #5155: Fix expressions in Database
This task is automatically imported from the old Task Issue Board and it was originally created by Radosław Waśko.
Original issue is [here](https://www.pivotaltracker.com/story/show/183933832).
<hr />

Expression tests were only being run for the in-memory backend.

Turns out many of them fail on the Database backend. They have been marked as pending for now, but the issues should be fixed so that the expressions work properly on either backend.

Review all the Column operations and consistency between InMemory and InDB operation
- Pay attention to NULL in equality (both should allow for a == Null).
- Division in SQL appears to be integer based if both columns are integer.

## Comments:
**James Dunkerley** reports a new **STANDUP** for today (2022-12-06):

**Progress:** Covering the PR tidying up. Build an experimental flow downloading the Nasdaq price history and processing. Adding Data.fetch API prototype. Started looking at the issues with Database expressions. It should be finished by 2022-12-12.

**Next Day:** Next day I will be working on the [#183933832 task](https://www.pivotaltracker.com/story/show/183933832). Finish Data.fetch prototype. Continue the expression debugging. (Enso Bot - Dec 6, 2022)
<hr />
Missing loads of testing of the column operations - this will determine follow on tickets. (James Dunkerley - Dec 20, 2022)
<hr />

### PR DESCRIPTION
**Pull request by jdunkerley**
### Pull Request Description

- Adjust Excel Workbook write behaviour.
- Support Nothing / Null constants.
- Deduce the type of arithmetic operations and `iif`.
- Allow Date_Time constants, treating as local timezone.
- Removed the `to_column_name` and `ensure_sane_name` code.

### Checklist

Please include the following checklist in your PR:

- [x] The documentation has been updated if necessary.
- [x] All code conforms to the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides.
- All code has been tested:
  - [x] Unit tests have been written where possible.
  - [ ] If GUI codebase was changed: Enso GUI was tested when built using BOTH
        `./run ide build` and `./run ide watch`.

### CODE REVIEW

**Review by GregoryTravis — APPROVED** [review]

File: distribution/lib/Standard/Database/0.0.0-dev/src/Data/Column.enso [hunk 1]
```diff
@@ -537,8 +546,10 @@ type Column
            common type.
         left_type = get_approximate_type when_true self.sql_type
         right_type = get_approximate_type when_false self.sql_type
-        if left_type != right_type then Error.throw (Illegal_Argument.Error "when_true and when_false types do not match") else
-            self.make_op "IIF" [when_true, when_false] new_type=left_type
+        new_type = SQL_Type.arithmetic_type left_type right_type
```
Comment (radeusgd) [inline distribution/lib/Standard/Database/0.0.0-dev/src/Data/Column.enso]: Feels like slightly abusing the `arithmetic_type` function. But I think it's ok given we are planning to revise this soon. And it works.

File: distribution/lib/Standard/Database/0.0.0-dev/src/Internal/JDBC_Connection.enso [hunk 2]
```diff
@@ -179,7 +179,10 @@ set_statement_values stmt holes =
     holes.map_with_index ix-> obj->
         position = ix + 1
         case obj.first of
-            Nothing -> stmt.setNull position obj.second.typeid
+            Nothing ->
+                sql_type = if obj.second == SQL_Type.null then SQL_Type.varchar else obj.second
```
Comment (radeusgd) [inline distribution/lib/Standard/Database/0.0.0-dev/src/Internal/JDBC_Connection.enso]: Huh? I'd appreciate a comment why this is happening.
I understand that we treat a value from a Null column as text for some reason? Why?

File: distribution/lib/Standard/Table/0.0.0-dev/src/Data/Expression.enso [hunk 3]
```diff
@@ -32,7 +32,9 @@ type Expression
        Converts an expression into a compatible column name.
     to_column_name : Text -> Text
     to_column_name expression =
-        expression.replace "[^A-Za-z_0-9]" "_" matcher=Regex_Matcher.Value
+        no_returns = expression.replace '\r' " " matcher=Text_Matcher.Case_Sensitive
+        no_new_lines = no_returns.replace '\n' " " matcher=Text_Matcher.Case_Sensitive
+        no_new_lines.replace "[^A-Za-z_0-9]" "_" matcher=Regex_Matcher.Value
```
Comment (radeusgd) [inline distribution/lib/Standard/Table/0.0.0-dev/src/Data/Expression.enso]: I'm a bit confused. This function is modified, but then also it seems it is not used at all in the codebase. Shall we just remove it?

File: std-bits/database/src/main/java/org/enso/database/JDBCProxy.java [hunk 4]
```diff
@@ -37,4 +39,14 @@ public static Object[] getDrivers() {
...
+
+  /**
+   * Converts a ZonedDateTime to a Timestamp (note loses the timezone).
+   *
+   * @param zonedDateTime the ZonedDateTime to convert
+   * @return the converted Timestamp
+   */
+  public static Timestamp getTimestamp(ZonedDateTime zonedDateTime) {
+    return Timestamp.from(zonedDateTime.toInstant());
+  }
```
Comment (radeusgd) [inline std-bits/database/src/main/java/org/enso/database/JDBCProxy.java]: `JDBCProxy` is a class created specifically to initialize JDBC drivers/connections avoiding the classloader issues from within the language.

I'm not sure if this other helper fits it. Maybe we can move it to a separate class next to this one? (For example maybe we could create `JDBCUtils.java` in the same package)

**Review by radeusgd — APPROVED** [review]
Looks good, just a few suggestions.

Btw. does the removal of `to_column_name` usages mean that this basically solves #5583 or do we still need to verify something? I guess before we close that one we should add a few 'adversarial' examples that verify that the escaping works correctly, but on the implementation level it seems to be mostly complete?

File: distribution/lib/Standard/Table/0.0.0-dev/src/Data/Expression.enso (hunk 3 above)
Comment (jdunkerley) [reply distribution/lib/Standard/Table/0.0.0-dev/src/Data/Expression.enso]: Yes meant to remove it - it appears we are fine will strange and wonderful names :)

Comment (radeusgd) [reply distribution/lib/Standard/Table/0.0.0-dev/src/Data/Expression.enso]: I think I want to try putting emojis there and see SQLite crying 😉

**Pull request merged**